core/vm/evm.go

```diff
@@ -184,14 +189,44 @@ func (evm *EVM) call(typ OpCode, caller ContractRef, addr libcommon.Address, inp
 	if evm.config.NoRecursion && evm.depth > 0 {
 		return nil, gas, nil
 	}
+	// CS TODO: need to check whether we should start call here.
+	if evm.firehoseContext.Enabled() {
+		typStr := typ.String()
+		callerAddress := caller.Address()
+
+		switch typ {
+		case STATICCALL:
+			typStr = "STATIC"
+		case DELEGATECALL:
+			typStr = "DELEGATE"
+			// Firehose a Delegate Call is quite different then a standard Call or event Call Code
+			// because it executes using the state of the parent call. Assumuming a contract that
+			// receives a method `execute`, let's say this contract is A. When in the `execute`
+			// method a `delegatecall` is performed to contract B, the net effect is that code of
+			// B is loaded and executed against the current state and value of contract A. As such,
+			// the real caller is the one that called contract A.
+			// It's a sure thing that caller is a Contract, it cannot be anything else, so we are safe
+			parent := caller.(*Contract)
+			callerAddress = parent.Address()
+		}
+		evm.firehoseContext.StartCall(typStr)
+		evm.firehoseContext.RecordCallParams(typStr, callerAddress, addr, value, gas, input)
```

Comment:
The value in case of `staticCall` is `firehose.EmptyValue`. It's better to have code as mentioned above.